Directory management on Local (Project Manager) backend

somebody1234 opened the pull request:
### Pull Request Description
- Implement https://github.com/enso-org/cloud-v2/issues/961
  - Allow directory management on Local backend
    - Setting a custom root directory is currently *out of scope* of this PR.
    - Listing directories
    - Deleting directories and files
    - Adjust project-related APIs to accept parent directory path (as required by PM when not interacting with root directory)
- QoL improvements related to testing this PR
  - New watch script (`npm run watch2`, `npm run watch:linux`) in `app/ide-desktop/lib/client/`) for testing IDE2 on Electron without having to build the entire app
  - Adjustments to `gui2`'s `vite.config.ts` to allow React HMR when doing dev in Electron

### Important Notes
- Support for deleting files and folders uses the API introduced by #9359 - so it will not work until that PR is merged in.
- Support for uploading files uses the API specified by #9360 - so it will not work until that issue is closed.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] ~~The documentation has been updated, if necessary.~~
- [x] ~~Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.~~
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] ~~Unit tests have been written where possible.~~
  - [x] ~~If GUI codebase was changed, the GUI was tested when built using `./run ide build`.~~

somebody1234 committed:
WIP: Add PM FS API

Stuck on distinguishing between files, folders and projects

somebody1234 committed:
Use PM for listing directories

somebody1234 committed:
Remove distinction between Local and Remote backends when listing directories

somebody1234 committed:
Add `projectsDirectory` to the appropriate PM API wrappers

somebody1234 committed:
Fix listing directory

somebody1234 committed:
Merge branch 'develop' into wip/sb/pm-directory-management

somebody1234 committed:
Initial integration of PM directory listing; add new watch script for `ide-desktop/lib/client`

somebody1234 committed:
Fix type errors

somebody1234 committed:
Fix default directory on Local backend; attempt to make opening projects work

somebody1234 committed:
Use Vite dev server for `watch2` script in `app/ide-desktop/lib/client/`

somebody1234 committed:
Fix lint and type errors

somebody1234 committed:
Add convenience Linux variant of Electron watch script

somebody1234 committed:
Minor QoL adjustments to watch functionality

somebody1234 committed:
Fix error when listing directories on Local backend

somebody1234 committed:
Add support for deleting directories

somebody1234 committed:
Enable button for creating folder in Local backend; fix navigation prevention

somebody1234 committed:
WIP: Fix HMR for dashboard when run through GUI2

somebody1234 committed:
Implement and test "Add folder" on Local backend

somebody1234 committed:
Enable `New Project` option; fix creating new project inside folder when on Local backend

TODO: Pass `parentDirectory` to the Rename and Delete endpoints

somebody1234 committed:
Pass `parentId` to `createProject` and `updateProject` on Local backend

somebody1234 committed:
Fixes

somebody1234 committed:
Only allow downloading projects on Local backend

somebody1234 committed:
Implement moving files

somebody1234 committed:
Allow dragging files on Local backend

somebody1234 committed:
Update `path` on project metadata on move

somebody1234 committed:
Make dragging an item into its existing directory a no-op

somebody1234 committed:
Allow deleting files

somebody1234 committed:
Attempt to fix `driveView.spec.ts` flakiness

somebody1234 committed:
Refactor PM command endpoint to use URL search params; add `uploadFile` endpoint to Local backend

somebody1234 committed:
Remove restriction allowing only projects to be uploaded on Local Backend

somebody1234 committed:
Merge branch 'develop' into wip/sb/pm-directory-management

Frizi reviewed (approved):
Approving gui vite config changes.

somebody1234 committed:
Merge branch 'develop' into wip/sb/pm-directory-management

somebody1234 commented:
hmm. i suppose cut & paste functionality should also be enabled for the Local backend, given that drag-n-drop is enabled

indiv0 commented:
❌ QA failed

- Listing directories (not projects, that one is present in `project/list`), indicating files, directories and projects
    - ❌ See **"Subdirectory move bug"**
- ✅ Creating directories
- Creating files
    - 🔶 ~~See **"Upload file weird default"** for potential issue~~ (Out of scope until https://github.com/enso-org/enso/issues/9360)
    - ❌ See **"Can't download project"**
    - ❌ See **"Can't upload project from Cloud"**
- ✅ Deleting directories
- ✅ Deleting files
- ✅ Detecting projects in directories (a smarter version of `project/list` that scans the provided directory only)
- Support for moving files to some target destination
    - :x: See **"File move bug"**
- Support for moving directories to some target destination
    - :x: Duplicate: see **"Subdirectory move bug"**
- 🔶 ~~Setting a custom root directory~~ (out of scope)
- Adjust project-related APIs to accept parent directory path (as required by PM when not interacting with root directory)
    - Create a subdirectory (in the local backend), then open/close/delete a project in that subdirectory
        - :x: See **"Project subdirectory bug"**
- Adjustments to `gui2`'s `vite.config.ts` to allow React HMR when doing dev in Electron
    - ✅ New watch script rebuilds IDE2 on Electron on code changes
        - `cd app/ide-desktop/lib/client && npm run watch2`
        - Edit any code in the `dashboard/app`, some text that you can visibly see
        - ⚠️ See **"`watch2` missing defaults"**
        - ❌ See **"`watch2` HMR doesn't work"**
        - :x: See **"`watch2` creates `~` directory"**
    - ⚠️ New watch script rebuilds IDE2 on Electron on code changes
        - `cd app/ide-desktop/lib/client && npm run watch:linux`
        - Edit any code in the `dashboard/app`, some text that you can visibly see
        - **No Linux machine available, someone else should QA this**
        
<details>
<summary>
<b>Subdirectory move bug</b>
</summary>

Steps to reproduce:

1. Create directory `A` under root

2. Open directory `A`

3. Create directory `B` under root

4. Move directory `B` into directory `A`

5. Open directory `B`

6. Observe error in logs and infinite spinner

https://github.com/enso-org/enso/assets/641547/2915ff3f-b9d6-4fb4-a3f2-3be1fd033a75
</details>


<details>
<summary>
<b>File move bug</b>
</summary>

Steps to reproduce:

1. Create directory `A` under root

2. Create directory `B` under `A`

3. Create file `test.csv` under root (had to do this manually by going to `~/enso/projects`)

4. Refresh file list to observe the file

5. Move `test.csv` to directory `B`

6. Move `test.csv` to root

7. Observe error message

https://github.com/enso-org/enso/assets/641547/59dc827e-5cbc-4b2a-8f4d-67a77954347a
</details>

<details>
<summary>
<b>Upload file weird default</b>
</summary>

When trying to upload a file, it only lets you upload `.enso-project` by default.

Is this behaviour that we still want? Users probably expect to be able to upload any kind of file into Enso now (e.g., `.csv`) and making them take an extra step to upload it seems like extra friction.

<img width="1470" alt="image" src="https://github.com/enso-org/enso/assets/641547/4796ec3b-2756-48c5-a5be-f269e81cf7bc">
</details>

<details>
<summary>
<b>Can't download project</b>
</summary>

When trying to download an `.enso-project` from the local projects browser, nothing happens.

Steps to reproduce:

1. Create a project

2. Click the download project button

3. Select a location to save the project to

4. Observe that no project is saved to that location


https://github.com/enso-org/enso/assets/641547/2a55ebc4-dedf-4156-a351-2f90017b0de5
</details>

<details>
<summary>
<b>Can't upload file from Cloud</b>
</summary>

When trying to upload an `.enso-project` downloaded from the Cloud S3, an error message is shown.

Steps to reproduce:

1. Download a `.enso-project` from the Cloud S3 (one is attached to this issue as a .zip -- extract it first)
2. Click the upload project button in local project manager
3. Select the `.enso-project` for upload
4. Observe the error

https://github.com/enso-org/enso/assets/641547/8612d92b-1ce8-45d7-b980-976a76754645

[project_root.enso-project.zip](https://github.com/enso-org/enso/files/14606298/project_root.enso-project.zip)
</details>


<details>
<summary>
<b>Project subdirectory bug</b>
</summary>

Steps to reproduce:

1. Create directory `A` under root

2. Create directory `B` under `A`

3. Create new project under `B`

4. Observe error message


https://github.com/enso-org/enso/assets/641547/440e8f3d-a33c-42d4-98ef-e6ab321412e0
</details>

<details>
<summary>
<b>`watch2` missing defaults</b>
</summary>

Wasn't able to get `watch2` to work by default. Had to set environment variables manually to get it to run. Apparently this is because the environment variables aren't cross-platform. We should still find a way to set reasonable defaults, or introduce a `npm run watch2:macos`.

I got it to work with:

```shell
(export ENSO_GIT=$(pwd) && \
    cd app/ide-desktop/lib/client && \
	ENSO_CLOUD_ENVIRONMENT=npekin \
	ENSO_BUILD_GUI=${ENSO_GIT}/dist/gui \
	ENSO_BUILD_ICONS=/tmp/.tmpNb8f1E \
	ENSO_BUILD_IDE=${ENSO_GIT}/dist/ide \
	ENSO_BUILD_PROJECT_MANAGER=${ENSO_GIT}/dist/backend \
	ENSO_BUILD_PROJECT_MANAGER_IN_BUNDLE_PATH=${ENSO_BUILD_PROJECT_MANAGER_IN_BUNDLE_PATH:-bin/project-manager} \
	ENSO_BUILD_IDE_BUNDLED_ENGINE_VERSION=${ENSO_BUILD_IDE_BUNDLED_ENGINE_VERSION:-0} \
	npm run watch2)
```
</details>

<details>
<summary>
<b>`watch2` HMR doesn't work</b>
</summary>

`watch2` successfully rebuilds the IDE on code changes, but HMR doesn't seem to be enabled. With network throttling enabled, a visible full reload occurs.

https://github.com/enso-org/enso/assets/641547/87fc97b2-cb51-451c-acf2-2827a59e44fd
</details>

<details>
<summary>
<b>`watch2` creates `~` directory</b>
</summary>

When running `watch2`, I ended up with a directory called `~` in my `/Users/indiv0/src/enso` directory. This seems like unintentional behaviour, and potentially spooky since a wrongly typed `rm -rf` to clean it up could do some damage.

<img width="465" alt="image" src="https://github.com/enso-org/enso/assets/641547/439bcb27-a161-4c65-8b09-ca5493590bae">
</details>

somebody1234 committed:
Enable "cut" and "cut all" on Local backend

somebody1234 committed:
Remove `.enso-project` restriction for uploading files on Local backend

somebody1234 committed:
Move `watch2` to `watch`; add `watch:macos`

somebody1234 committed:
Switch watch scripts to use `/tmp`

somebody1234 committed:
Fix local project download not working

indiv0 commented on app/ide-desktop/lib/client/esbuild-config.ts:
```diff
@@ -55,9 +57,13 @@ export function bundlerOptions(
             BUNDLED_ENGINE_VERSION: JSON.stringify(bundledEngineVersion),
             PROJECT_MANAGER_IN_BUNDLE_PATH: JSON.stringify(projectManagerInBundlePath),
+            'process.env.ELECTRON_DEV_MODE': JSON.stringify(String(devMode)),
+            'process.env.GUI2_CONFIG_PATH': JSON.stringify(
```
Should this env var be renamed to `GUI_CONFIG_PATH` now that `GUI2` is default?

somebody1234 replied:
yeah - nice catch

indiv0 commented on app/ide-desktop/lib/client/src/bin/server.ts:
```diff
@@ -125,13 +133,27 @@ export class Server {
 
-    /** Respond to an incoming request. */
+    /** Respond to an incoming request.
+     * @throws {Error} on malformed input. */
```
Might be helpful to explain what `malformed input` means in this case, or provide an example

indiv0 commented on app/ide-desktop/lib/client/src/bin/server.ts:
```diff
@@ -176,22 +198,62 @@ export class Server {
+                        () => {
+                            response
+                                .writeHead(HTTP_STATUS_BAD_REQUEST, common.COOP_COEP_CORP_HEADERS)
+                                .end()
```
Should we include some information on what failed here?

somebody1234 replied:
i don't think so - i don't think we have more specific information, so the default "<action x> failed" message that the frontend outputs for most errors should be sufficient

indiv0 commented on app/ide-desktop/lib/client/src/bin/server.ts:
```diff
@@ -176,22 +198,62 @@ export class Server {
-                            .end(id)
-                    })
+                            .writeHead(HTTP_STATUS_BAD_REQUEST, common.COOP_COEP_CORP_HEADERS)
+                            .end('Command arguments must be an array of strings')
```
Error messages should contain trailing period, right?

indiv0 commented on app/ide-desktop/lib/client/src/index.ts:
```diff
@@ -419,6 +422,24 @@ class App {
             address.port = this.serverPort().toString()
             address.search = new URLSearchParams(searchParams).toString()
             logger.log(`Loading the window address '${address.toString()}'.`)
+            if (process.env.ELECTRON_DEV_MODE === 'true') {
+                const window = this.window
+                const onLoad = () => {
+                    void window.webContents.mainFrame
+                        .executeJavaScript('document.body.innerHTML')
+                        .then(html => {
+                            if (html === '') {
+                                console.warn('Loading failed, reloading...')
+                                window.webContents.once('did-finish-load', onLoad)
+                                setTimeout(() => {
+                                    void window.loadURL(address.toString())
+                                    // eslint-disable-next-line @typescript-eslint/no-magic-numbers
+                                }, 1_000)
+                            }
+                        })
+                }
+                window.webContents.once('did-finish-load', onLoad)
+            }
```
Would be nice to have some comments to explain what's going on here

indiv0 commented on app/ide-desktop/lib/client/watch.ts:
```diff
@@ -41,14 +37,20 @@ const PROJECT_MANAGER_BUNDLE_PATH = paths.getProjectManagerBundlePath()
     void (async () => {
         console.log('Bundling client.')
-        const clientBundlerOpts = clientBundler.bundlerOptionsFromEnv()
+        const clientBundlerOpts = clientBundler.bundlerOptionsFromEnv(true)
```
Could use a constant explaining what this `true` represents

indiv0 commented on app/ide-desktop/lib/client/watch.ts:
```diff
@@ -72,48 +74,7 @@ const ALL_BUNDLES_READY = new Promise<Watches>((resolve, reject) => {
-        console.log('Bundling dashboard.')
-        const dashboardOpts = dashboardBundler.bundleOptions()
-        dashboardOpts.plugins.push({
-            name: 'enso-on-rebuild',
-            setup: build => {
-                build.onEnd(() => {
-                    console.log('Dashboard bundle updated.')
-                })
-            },
-        })
-        dashboardOpts.outdir = path.resolve(IDE_DIR_PATH, 'assets')
-        const dashboardBuilder = await esbuild.context(dashboardOpts)
-        const dashboard = await dashboardBuilder.rebuild()
-        console.log('Result of dashboard bundling: ', dashboard)
-        // We do not need to serve the dashboard as it outputs to the same directory.
-        // It will not rebuild on request, but it is not intended to rebuild on request anyway.
-        // This MUST be called before `builder.watch()` as `tailwind.css` must be generated
-        // before the copy plugin runs.
-        void dashboardBuilder.watch()
-
-        console.log('Bundling content.')
-        const contentOpts = contentBundler.bundlerOptionsFromEnv({
-            supportsLocalBackend: true,
-            supportsDeepLinks: false,
-        })
-        contentOpts.plugins.push({
-            name: 'enso-on-rebuild',
-            setup: build => {
-                build.onEnd(() => {
-                    console.log('Content bundle updated.')
-                })
-            },
-        })
-        contentOpts.pure.splice(contentOpts.pure.indexOf('assert'), 1)
-        contentOpts.outdir = path.resolve(IDE_DIR_PATH, 'assets')
-        contentOpts.define.REDIRECT_OVERRIDE = JSON.stringify('http://localhost:8080')
-        const contentBuilder = await esbuild.context(contentOpts)
-        const content = await contentBuilder.rebuild()
-        console.log('Result of content bundling: ', content)
-        void contentBuilder.watch()
```
How come we don't need this anymore? Do we not need to rebuild the content and dashboard bundles?

somebody1234 replied:
correct:
- `content/` was the entrypoint for GUI1
- `dashboard/` (as well as GUI2) are now served by GUI2's dev server (the Vite server) which rebuilds with HMR when possible

indiv0 commented on app/ide-desktop/lib/client/watch.ts:
```diff
@@ -129,10 +90,12 @@ const ELECTRON_ARGS = [path.join(IDE_DIR_PATH, 'index.cjs'), '--', ...process.ar
-    // from ending. Thus, we exit manually. It seems to terminate the child `esbuild` process
-    // as well.
-    process.exit(0)
+    void fs.rm(IDE_DIR_PATH, { recursive: true, force: true }).then(() => {
+        // The `esbuild` process seems to remain alive at this point and will keep our process
+        // from ending. Thus, we exit manually. It seems to terminate the child `esbuild` process
+        // as well.
+        process.exit(0)
+    })
```
Why do we delete the directory now? Don't we want to keep it around for future invocations of this script?

somebody1234 replied:
this is a watch script so it's probably marginally better to be sure that stray files aren't accidentally making the watch script work when actually we've deleted some very important files.

plus we use esbuild for bundling the electron app so re-building all the files should be fast enough (<<1s)

somebody1234 committed:
Fix error when moving subdirectories and files

indiv0 commented on app/ide-desktop/lib/dashboard/src/components/dashboard/ProjectNameColumn.tsx:
```diff
@@ -125,7 +125,7 @@ export default function ProjectNameColumn(props: ProjectNameColumnProps) {
           try {
             const createdProject = await backend.createProject({
-              parentDirectoryId: asset.parentId,
+              parentId: asset.parentId,
```
Why the rename from `parentDirectoryId` to `parentId`?

somebody1234 replied:
for consistency... but this is a bug since it's used verbatim in the payload to the Cloud backend. nice catch, thanks!

somebody1234 committed:
Fix error toasts

indiv0 commented on app/ide-desktop/lib/dashboard/src/services/Backend.ts:
```diff
@@ -972,7 +987,17 @@ export function compareAssets(a: AnyAsset, b: AnyAsset) {
     return relativeTypeOrder
   }
-  return a.title > b.title ? 1 : a.title < b.title ? COMPARE_LESS_THAN : 0
+  const aModified = new Date(a.modifiedAt)
+  const bModified = new Date(b.modifiedAt)
+  return aModified > bModified
+    ? -1
+    : aModified < bModified
+      ? 1
+      : a.title > b.title
+        ? 1
+        : a.title < b.title
+          ? -1
+          : 0
```
😱😱😱

Generally I appreciate terseness but nested ternary operators are a huge pain to read. Could we make this more readable? Even an if-pyramid would be fine here.

indiv0 commented on app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts:
```diff
@@ -5,13 +5,82 @@
 import ProjectManager from '#/utilities/ProjectManager'
 
+// =============
+// === Types ===
```
My understanding of the style guide is that headers should match the name of the item they precede. So we should have a `ProjectMetadata` header, an `Attributes` header, etc.

They're not meant to denote "sections" of the code, but to help find items on a quick scroll.

somebody1234 replied:
hm. you might be right...
<https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md#sections>
i think it's out of scope for this PR though, since i've been using headers for `Types` and `Constants` pretty liberally so it'd require changes in a lot of other places.

also one issue with that approach, is that in the dashboard, it's more or less one "concept"/main structure per file. most of the exceptions are in utility modules that export a bunch of functions instead of a class
so that means these "sections" would be kinda redundant

indiv0 commented on app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts:
```diff
@@ -21,6 +90,66 @@ function ipWithSocketToAddress(ipWithSocket: projectManager.IpWithSocket) {
   return backend.Address(`ws://${ipWithSocket.host}:${ipWithSocket.port}`)
 }
 
+// ======================================
+// === Functions for manipulating ids ===
+// ======================================
```
ibid. re: header naming

indiv0 commented on app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts:
```diff
@@ -21,6 +90,66 @@ function ipWithSocketToAddress(ipWithSocket: projectManager.IpWithSocket) {
+// === Functions for manipulating ids ===
+// ======================================
+
+/** Create a {@link backend.DirectoryId} from a path. */
```
Just to help me understand -- what is a `path` in this case? Is it an actual path on the filesystem? If it is, perhaps `Path` should be typed?

indiv0 commented on app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts:
```diff
@@ -21,6 +90,66 @@ function ipWithSocketToAddress(ipWithSocket: projectManager.IpWithSocket) {
+  const [, typeRaw, idRaw] = id.match(/(.+?)-(.+)/) ?? []
+  switch (typeRaw) {
+    case backend.AssetType.directory: {
+      return { type: backend.AssetType.directory, id: backend.DirectoryId(idRaw ?? '') }
```
How come `id` can be an empty string?

somebody1234 replied:
it shouldn't be able to, but we cannot statically assert that even if we wanted to, because TS doesn't parse regexes and so we don't know the number of matches at the type level

indiv0 commented on app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts:
```diff
@@ -44,29 +175,73 @@ export default class LocalBackend extends Backend {
     }
   }
 
+  /** Return the ID of the root directory. */
+  override rootDirectoryId(): backend.DirectoryId | null {
+    return newDirectoryId(this.defaultRootDirectory)
+  }
```
Are we sure we want to allow this to return `null`? How can that happen when we're dealing with the local backend?

somebody1234 replied:
the types were copy-pasted by VSCode from `Backend`. it's true that this shouldn't ever return `null`.

indiv0 commented on app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts:
```diff
@@ -44,29 +175,73 @@ export default class LocalBackend extends Backend {
+              parentId,
+              permissions: [],
+              projectState: {
+                type: LocalBackend.currentlyOpenProjects.has(entry.metadata.id)
+                  ? backend.ProjectState.opened
+                  : entry.metadata.id === LocalBackend.currentlyOpeningProjectId
+                    ? backend.ProjectState.openInProgress
+                    : backend.ProjectState.closed,
```
It seems like having a `LocalBackend.currentlyOpenProjects` results in two sources of truth for the state of the projects. Why not have `LocalBackend.projects` and let the projects take care of their own state?

somebody1234 replied:
this awkward structure is due to only the `openProject` endpoint being the only one with enough info to construct the full project info...

... but i don't see any downsides of combining the two properties anyway, so i guess i'll do it

indiv0 replied:
If you want, you can open an issue describing which backend endpoints you'd like modified to provide the necessary info. The backend's job is to make the frontend's job easier ;)

indiv0 commented on app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts:
```diff
@@ -271,25 +457,35 @@ export default class LocalBackend extends Backend {
+        await this.runProjectManagerCommand(null, 'filesystem-delete', typeAndId.id)
+        return
+      }
+      case backend.AssetType.project: {
+        if (LocalBackend.currentlyOpeningProjectId === typeAndId.id) {
+          LocalBackend.currentlyOpeningProjectId = null
+        }
+        LocalBackend.currentlyOpenProjects.delete(newProjectId(typeAndId.id))
+        try {
+          await this.projectManager.deleteProject({
+            projectId: typeAndId.id,
+            projectsDirectory: extractTypeAndId(body.parentId).id,
+          })
+          return
+        } catch (error) {
+          throw new Error(
+            `Could not delete project ${
+              title != null ? `'${title}'` : `with ID '${typeAndId.id}'`
+            }: ${errorModule.tryGetMessage(error) ?? 'unknown error'}.`
+          )
+        }
+      }
```
This is what I mean by two sources of truth.

If we had a `LocalBackend.projects` then `LocalBackend.currentlyOpeningProjectId` could be materialized by filtering over `LocalBackend.projects` and we wouldn't have to update it separately from `LocalBackend.currentlyOpenProjects`.

Setting `project.projectState = closed` on `LocalBackend.projects` would be sufficient to update both properties in that case.

indiv0 commented on app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts:
```diff
@@ -385,9 +581,55 @@ export default class LocalBackend extends Backend {
+        : extractTypeAndId(params.parentDirectoryId).id
+    const path = `${parentPath}/${params.fileName}`
+    await this.runProjectManagerCommand(file, 'filesystem-write-path', path)
+    // `project` MUST BE `null` as uploading projects uses a separate endpoint.
+    return { path, id: newFileId(path), project: null }
```
If it must be `null` then why is it a field at all? Is this because `remoteBackend` provides it? Perhaps `remoteBackend` should be extending the type in that case?

somebody1234 replied:
perhaps... it's possible that the current structure is partly historical baggage? in a nutshell:
- both `LocalBackend` and `RemoteBackend` are implementations of the abstract interface `Backend`. the impls aren't exposed directly, only a `Backend` of an arbitrary type
- but also the API has been designed using the `RemoteBackend` (cloud backend) as the source of truth - you can see that the `LocalBackend` does quite a bit of transformation on various endpoints to make the shape consisteent with the one returned by `RemoteBackend`

indiv0 commented on app/ide-desktop/lib/dashboard/src/services/LocalBackend.ts:
```diff
@@ -489,4 +720,47 @@ export default class LocalBackend extends Backend {
+    }
+  }
+
+  /** List directories, projects and files in the given folder. */
+  private async projectManagerFilesystemList(parentId: backend.DirectoryId | null) {
+    /** The type of the response body of this endpoint. */
+    interface ResponseBody {
+      readonly entries: FileSystemEntry[]
+    }
+
+    const response = await this.runProjectManagerCommand<ResponseBody>(
+      null,
+      'filesystem-list',
+      parentId ?? this.defaultRootDirectory
+    )
+    return response.entries
+  }
```
How come some projectManager-interacting methods are defined as private functions here, and some are defined like `projectManager.deleteAsset`. Seems like we should be all-in on one or the other, no?

somebody1234 replied:
one kind uses the existing WS connection, the other kind (this kind) goes through a HTTP proxy defined by Electron to `exec` the actual project manager binary - supposedly these endpoints are not exposed in WS for security purposes (the user invoking the command needs to have access to the files to be able to `exec` the binary)

somebody1234 replied:
i suppose they should be defined on the project manager class regardless...

indiv0 reviewed (changes requested):
Some comments and some suggestions.

Mostly questions to help my own understanding of what's going on, since its been a while since I've looked at the dashboard code.

somebody1234 committed:
Really fix error when moving subdirectories and files; improve error toasts

somebody1234 committed:
Fix interacting with projects in subdirectories

somebody1234 commented:
- Subdirectory move bug: fixed.
- File move bug: fixed.
- Project subdirectory bug: probably fixed?
- Upload file weird default: fixed.
- Can't download project: fixed.
- `watch2` (now `watch`) missing defaults: `watch:macos` added.

unrelated changes:
- "Cut" and "Cut All" shortcuts and context menu entries: enabled on Local backend.

remaining bugs:
- Can't upload file from Cloud: not yet investigated
- `watch2` HMR doesn't work: (!) will have to investigate. i do remember it working at some point, it's possible that it's broken since then
- `watch2` creates `~` directory: i know the cause, but i think it'll be a bit tricky to fix. i'll come back to it later.

somebody1234 committed:
Address review (round 1)

somebody1234 committed:
Fix types of Project Manager API

somebody1234 committed:
Move Project Manager CLI (i.e. not WS) endpoints to `ProjectManager.ts`

somebody1234 committed:
Combine `currentlyOpeningProjectId` and `currentlyOpenProjects` into `projects`

somebody1234 committed:
Merge branch 'develop' into wip/sb/pm-directory-management

somebody1234 commented:
okay two main problems on the backend side:
- it lacks a way to get the default root directory. the current solution is to use `~/enso/projects/` and get Electron to handle `~/`. it's far from an ideal solution, of course...
- i'm not sure whether the HTTP endpoint for downloading files was designed for nested paths (and even less sure about absolute paths...)

somebody1234 commented:
current status: *most* issues fixed (see above for full status)
all reviews should also be addressed now.

indiv0 reviewed (approved):
LGTM, will re-QA tomorrow

indiv0 commented:
> * Can't download project: fixed.

Potentially related to:
- https://github.com/enso-org/enso/pull/9297
- https://github.com/enso-org/cloud-v2/issues/947

somebody1234 commented:
> Potentially related to:

nah. those are for cloud - the issue here is that i was using the wrong id - `project-<uuid>` instead of just plain `<uuid>`. so it should be a regression that was specific to the PR, as this PR is the one that introduced `project-<uuid>` format ids in the first place - i assume it still works fine on develop.

somebody1234 committed:
Fix root directory of Local backend

somebody1234 committed:
Move `ProjectManager.ts` to a more appropriate place

somebody1234 committed:
Fix incorrect root directory

somebody1234 committed:
Partially fix issue with uploaing file from Cloud

somebody1234 commented:
- `watch2` creates `~` directory: should be fixed.
- Can't upload file from Cloud: *partially* fixed. still errors because we should no longer be using `listProjects` which does not support specifying the parent directory.
  - the more pressing problem, however, is that the project upload code is written in the Electron app. i'll ask whether this functionality can be added to the PM instead - that way we could potentially even get back metadata
- `watch2` HMR doesn't work: can repro... that's weird.

somebody1234 committed:
Fix HMR?

indiv0 commented:
> Can't upload file from Cloud: _partially_ fixed. still errors because we should no longer be using `listProjects` which does not support specifying the parent directory.
> 
>     * the more pressing problem, however, is that the project upload code is written in the Electron app. i'll ask whether this functionality can be added to the PM instead - that way we could potentially even get back metadata

Per discussion on discord:

> the current upload functionality is implemented in the Electron app rather than the PM. i wanted to use this as an opportunity to move the logic to the PM where it probably belongs better, but seems like that would take quite a but of effort on the PM's side

So won't be fixed in this PR or likely anytime soon, so ignoring for QA

somebody1234 committed:
Fix incorrect import path

somebody1234 commented:
correction: i'll probably adjust the "upload project" functionality on the Electron side to accept arbitrary paths - we shouldn't leave it broken for an indeterminate period of time, after all - it's just that i'll probably only get around to it on monday.

somebody1234 commented:
(i forgot to mention that the other issue is that uploading on the Local backend currently does not take a path... that definitely needs to be fixed too.)

indiv0 commented on app/gui2/src/components/GraphEditor.vue:
```diff
@@ -36,7 +36,8 @@ import { Vec2 } from '@/util/data/vec2'
 import * as set from 'lib0/set'
 import { computed, onMounted, ref, watch } from 'vue'
-import { ProjectManagerEvents } from '../../../ide-desktop/lib/dashboard/src/utilities/ProjectManager'
+// FIXME: Refactor this out to a shared dependency.
```
FIXMEs should be tracked with an issue

indiv0 commented:
:x: QA failed


- ✅ Listing directories
    - ✅ Subdirectory move bug
- ✅ Creating directories
- ❌ Creating files
    - ✅ Upload file weird default
    - :x: See **Upload file bug 2**
    - ✅ Can't download project
    - :x: Can't upload project from Cloud (wontfix?)
- ✅  Deleting directories
- ✅  Deleting files
- ✅  Detecting projects in directories
- ✅  Support for moving files to some target destination
    - ✅  File move bug
- ✅  Support for moving directories to some target destination
- ❌ Create a subdirectory (in the local backend), then open/close/delete a project in that subdirectory
    - ✅ Works when creating the project via context menu entry
    - :x: See **Project subdirectory bug 2**
    - :x: See **Project subdirectory bug 3**
- ✅  (MacOS) New watch script rebuilds IDE2 on Electron on code changes
        - ✅  `watch:macos` missing defaults
        - ✅  `watch:macos` HMR doesn't work (works now???)
        - ✅  `watch2` creates `~` directory
- ⚠️ (Linux) New watch script rebuilds IDE2 on Electron on code changes
    - **No Linux machine available, someone else should QA this**
- ✅  "Cut" and "Cut All" shortcuts and context menu entries: enabled on Local backend.
    - ⚠️ Works, but it doesn't seem to be possible to paste to the root directory -- can't open context menu on the root directory and `Cmd + V` doesn't work with no directory selected

<details>
<summary><b>HMR seems to work now</b></summary>

<img width="1470" alt="image" src="https://github.com/enso-org/enso/assets/641547/798531ba-e065-4e00-8b17-71cdec849049">

</details>


<details>
<summary><b>Upload file bug 2</b></summary>

I still see the expected error when attempting to upload a project, but I see a new type of error when trying to upload this log file. Not sure if it's to be expected or not.

https://github.com/enso-org/enso/assets/641547/fa63709d-e20d-4de0-ba66-d8611db725a8

[localhost-1710459847125.log](https://github.com/enso-org/enso/files/14623197/localhost-1710459847125.log)

</details>

<details>
<summary><b>Project subdirectory bug 2</b></summary>

When creating a project in a subdirectory, the project is created in the root directory instead.

Steps to reproduce:

1. Create a new directory

2. Select that directory

3. Click new project

4. Observe that project is not created in the subdirectory

https://github.com/enso-org/enso/assets/641547/b90e70ad-6aa2-4218-b2f7-1d3b33bb5280


</details>

<details>
<summary><b>Project subdirectory bug 3</b></summary>

When moving a newly created project to a subdirectory, an error appears.

Interestingly, if you refresh the page after the error, the project can be moved just fine.

Steps to reproduce:

1. Create a new directory

2. Create a new project

3. Wait for the project to open

4. Close the project

5. Move the project to the new directory

6. Observe that error

https://github.com/enso-org/enso/assets/641547/c01828a7-4eaf-4f9c-873b-2b3d415ce2cd


</details>

somebody1234 commented:
upload file bug 2: *technically* intentional - the PM endpoint doesn't exist yet. realistically though i think it makes sense to handle it via Electron for the time being, until the PM endpoint exists

indiv0 commented:
> upload file bug 2: _technically_ intentional - the PM endpoint doesn't exist yet. realistically though i think it makes sense to handle it via Electron for the time being, until the PM endpoint exists

Would be better, because the current behaviour crashes the app.

somebody1234 commented:
> When creating a project in a subdirectory, the project is created in the root directory instead.

i think this is (was?) intentional. the shortcut and context menu both create it in the correct folder, the drive bar is meant for the root folder.

i suppose it does make sense to change this behavior though... after all, those buttons are *much* more discoverable than the context menu entries.

indiv0 commented:
- ❌ (Cloud) Open project on cloud fails with this branch

<details>
<summary>
<b>Open cloud project bug</b>
</summary>

Steps to reproduce:

1. Create a project on cloud dashboard

2. Click open project

3. Observe error in logs

Looks like this occurs because we're sending `parentId` to Cloud when we shouldn't be:

<img width="1470" alt="image" src="https://github.com/enso-org/enso/assets/641547/6d2b06fc-b147-40bc-abb2-62e89450a3e1">

</details>

somebody1234 committed:
Adjust "upload project" endpoint to accept parent directory path

somebody1234 committed:
Fix uploading projects to Local backend

somebody1234 committed:
Remove extra body parameter from Remote backend requests

somebody1234 committed:
Fix uploading files to Local backend

somebody1234 commented:
- Cannot upload project (from cloud or elsewhere): should be fixed
- Upload file bug 2: should be fixed
- Open cloud project bug: should be fixed
- Project subdirectory bug 2: should be fixed
- Project subdirectory bug 3: should be fixed

somebody1234 committed:
Fix moving newly created project

somebody1234 committed:
Make buttons add asset to subdirectories

somebody1234 committed:
Fix type error

somebody1234 commented:
@indiv0 *should* be ready for QA (again).

indiv0 commented:
:x:  QA failed

- ✅  Creating files
    - ✅ **Upload file bug 2** -- Can upload `.log` file
    - ~~Can't upload project from Cloud~~ (won't fix)
- :x: Create a subdirectory (in the local backend), then open/close/delete a project in that subdirectory
    - ✅ **Project subdirectory bug 2**
    - :x: See **Project subdirectory bug 3** (still broken)
    - :x: See **Project subdirectory bug 4**
- ⚠️ (Linux) New watch script rebuilds IDE2 on Electron on code changes
    - **No Linux machine available, someone else should QA this**
- ✅  "Cut" and "Cut All" shortcuts and context menu entries: enabled on Local backend.
    - ⚠️ Can't open context menu on root directory
    - ✅ Able to paste to the root directory by opening the context menu on a file in the root directory
    - Perhaps the "root directory area" needs to be expanded to include some space below the assets?
    - ⚠️ `Cmd + V` doesn't work to paste to root directory
- ✅ **(Cloud) Open project on cloud bug**


<details>
<summary><b>Project subdirectory bug 4</b></summary>

When moving a newly created project to the root directory, an error appears.


Steps to reproduce:

1. Create a new directory

2. Create a new project

3. Wait for the project to open

4. Close the project

5. Move the project to the root directory

6. Observe that error


https://github.com/enso-org/enso/assets/641547/4bf1b3f5-d918-4b16-b9dc-899242e574a1



</details>

somebody1234 committed:
Merge branch 'develop' into wip/sb/pm-directory-management

somebody1234 committed:
Fix project subdirectory bug 3?

indiv0 commented:
So what's the plan for this PR? As I understand it:
- fully fixing the bugs will require changes on the PM side
- not breaking `npm run dev` will require some unknown changes

The first one we can leave for a later PR, if we just disable the broken behaviour for now. Better disabled than erroring, I think

The latter is trickier. Not sure if we want to merge this without a way to avoid that breakage. Even a hack to make it work would be okay.

somebody1234 commented:
> fully fixing the bugs will require changes on the PM side

I think the way to go here is to wait for the PM changes:
- #9510

We *really* don't want to merge it in with this functionality disabled... because this affects creating *any* new project, whatsoever :scream: 

For the latter, the least cursed way might be to re-implement the HTTP API that's being used to proxy FS operations to the PM, in node itself - and use this equivalent node server in `npm run dev` in `app/gui2`. Not sure it's ideal, but it is *a* way to avoid the breakage.

somebody1234 committed:
Merge branch 'develop' into wip/sb/pm-directory-management

somebody1234 committed:
Initial implemementation of Project Manager shim

somebody1234 committed:
Fixes

somebody1234 committed:
Merge branch 'develop' into wip/sb/pm-directory-management

somebody1234 commented:
@indiv0 `npm run dev` on `app/gui2` should now work (and be testable).
note that as with before this PR, this *will* require a separate project manager process to be manually started: `$ENSO_GIT_ROOT/dist/backend/bin/project-manager`

somebody1234 committed:
Fix type error

somebody1234 committed:
Fix name of newly uploaded and created projects

somebody1234 committed:
Fix bug

indiv0 commented:
✅ QA passed

- ✅ `cd app/gui2 && ENSO_CLOUD_ENV=npekin npm run dev`
- ✅  Create a subdirectory (in the local backend), then open/close/delete a project in that subdirectory
    - ✅  **Project subdirectory bug 3**
    - ✅  **Project subdirectory bug 4**
- ⚠️ (Linux) New watch script rebuilds IDE2 on Electron on code changes
    - **No Linux machine available, someone else should QA this**
- ⚠️  "Cut" and "Cut All" shortcuts and context menu entries: enabled on Local backend.
    - ⚠️  Can open context menu on root directory, but doesn't show `Paste` option
    - ⚠️  `Cmd + V` doesn't work to paste to root directory

somebody1234 committed:
Merge branch 'develop' into wip/sb/pm-directory-management

somebody1234 committed:
Prettier
